src/torchcodec/decoders/_core/VideoDecoder.h

```diff
@@ -317,73 +304,117 @@ class VideoDecoder {
     bool operator!=(const DecodedFrameContext&);
   };
 
-  // Stores information for each stream.
   struct StreamInfo {
     int streamIndex = -1;
     AVStream* stream = nullptr;
     AVRational timeBase = {};
     UniqueAVCodecContext codecContext;
-    // The current position of the cursor in the stream.
+
+    // The FrameInfo indices we built when scanFileAndUpdateMetadataAndIndex was
+    // called.
+    std::vector<FrameInfo> keyFrames;
+    std::vector<FrameInfo> allFrames;
+
+    // The current position of the cursor in the stream, and associated frame
+    // duration.
     int64_t currentPts = 0;
     int64_t currentDuration = 0;
     // The desired position of the cursor in the stream. We send frames >=
     // this pts to the user when they request a frame.
-    // We update this field if the user requested a seek.
+    // We update this field if the user requested a seek. This typically
+    // corresponds to the decoder's desiredPts_ attribute.
     int64_t discardFramesBeforePts = INT64_MIN;
     VideoStreamOptions videoStreamOptions;
-    // The filter state associated with this stream (for video streams). The
-    // actual graph will be nullptr for inactive streams.
-    FilterState filterState;
+
+    // color-conversion fields. Only one of FilterGraphContextr and
+    // UniqueSwsContext should be non-null.
     ColorConversionLibrary colorConversionLibrary = FILTERGRAPH;
-    std::vector<FrameInfo> keyFrames;
-    std::vector<FrameInfo> allFrames;
-    DecodedFrameContext prevFrameContext;
+    FilterGraphContext filterGraphContext;
     UniqueSwsContext swsContext;
+
+    // Used to know whether a new FilterGraphContext or UniqueSwsContext should
+    // be created before decoding a new frame.
+    DecodedFrameContext prevFrameContext;
   };
 
-  // Returns the key frame index of the presentation timestamp using FFMPEG's
-  // index. Note that this index may be truncated for some files.
-  int getKeyFrameIndexForPtsUsingEncoderIndex(AVStream* stream, int64_t pts)
-      const;
-  // Returns the key frame index of the presentation timestamp using our index.
-  // We build this index by scanning the file in buildKeyFrameIndex().
-  int getKeyFrameIndexForPtsUsingScannedIndex(
-      const std::vector<VideoDecoder::FrameInfo>& keyFrames,
-      int64_t pts) const;
-  int getKeyFrameIndexForPts(const StreamInfo& stream, int64_t pts) const;
+  // --------------------------------------------------------------------------
+  // CONSTRUCTORS AND INITIALIZERS
+  // --------------------------------------------------------------------------
+  // Don't use those, use the static methods to create a decoder object.
+
+  explicit VideoDecoder(const std::string& videoFilePath, SeekMode seekMode);
+  explicit VideoDecoder(const void* buffer, size_t length, SeekMode seekMode);
```

Comment:
Flagging that per #477, we want to keep these public. We need to make sure that stays the case when we merge this PR with the results of #477 and #478.